executor/pkg_test.go

```diff
@@ -111,6 +117,102 @@ func (s *pkgTestSuite) TestNestedLoopApply(c *C) {
+	}
+	return exec
+}
+
+func prepare4RadixPartition(sctx sessionctx.Context, rowCount int) *HashJoinExec {
+	childExec0 := prepareOneColChildExec(sctx, rowCount)
+	childExec1 := prepareOneColChildExec(sctx, rowCount)
+
+	col0 := &expression.Column{Index: 0, RetType: types.NewFieldType(mysql.TypeLong)}
+	col1 := &expression.Column{Index: 0, RetType: types.NewFieldType(mysql.TypeLong)}
+	joinSchema := expression.NewSchema(col0, col1)
+	hashJoinExec := &HashJoinExec{
+		baseExecutor:   newBaseExecutor(sctx, joinSchema, "HashJoin", childExec0, childExec1),
+		concurrency:    4,
+		joinType:       0, // InnerJoin
+		innerIdx:       0,
+		innerKeys:      []*expression.Column{{Index: 0, RetType: types.NewFieldType(mysql.TypeLong)}},
+		innerKeyColIdx: []int{0},
+		outerKeys:      []*expression.Column{{Index: 0, RetType: types.NewFieldType(mysql.TypeLong)}},
+		outerKeyColIdx: []int{0},
+		innerExec:      childExec0,
+		outerExec:      childExec1,
+	}
+	return hashJoinExec
+}
+
+func (s *pkgTestSuite) TestRadixPartition(c *C) {
+	sctx := mock.NewContext()
+	hashJoinExec := prepare4RadixPartition(sctx, 200)
+	sv := sctx.GetSessionVars()
+	originL2CacheSize, originEnableRadixJoin, originMaxChunkSize := sv.L2CacheSize, sv.EnableRadixJoin, sv.MaxChunkSize
+	sv.L2CacheSize = 100
+	sv.EnableRadixJoin = true
+	sv.MaxChunkSize = 100
+	defer func() {
+		sv.L2CacheSize, sv.EnableRadixJoin, sv.MaxChunkSize = originL2CacheSize, originEnableRadixJoin, originMaxChunkSize
+	}()
+	sv.StmtCtx.MemTracker = memory.NewTracker("RootMemTracker", variable.DefTiDBMemQuotaHashJoin)
+
+	ctx := context.Background()
+	err := hashJoinExec.Open(ctx)
+	c.Assert(err, IsNil)
+
+	innerResultCh := make(chan *chunk.Chunk, hashJoinExec.concurrency)
+	doneCh := make(chan struct{})
+	go func() {
+		for range innerResultCh {
+		}
+	}()
+
+	hashJoinExec.fetchInnerRows(ctx, innerResultCh, doneCh)
+	c.Assert(hashJoinExec.innerResult.GetMemTracker().BytesConsumed(), Equals, int64(14400))
+
+	hashJoinExec.evalRadixBit()
+	// ceil(log_2(14400/(100*3/4))) = 8
+	c.Assert(len(hashJoinExec.innerParts), Equals, 256)
+	radixBits := hashJoinExec.radixBits
+	c.Assert(radixBits, Equals, uint32(0x00ff))
+
+	hashJoinExec.partitionInnerRows()
+	totalRowCnt := 0
+	for i, part := range hashJoinExec.innerParts {
+		if part == nil {
+			continue
+		}
+		totalRowCnt += part.NumRows()
+		iter := chunk.NewIterator4Chunk(part)
+		hasNull, keyBuf := false, make([]byte, 0, 64)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			hasNull, keyBuf, err = hashJoinExec.getJoinKeyFromChkRow(false, row, keyBuf)
+			c.Assert(err, IsNil)
+			c.Assert(hasNull, IsFalse)
+			joinHash := murmur3.Sum32(keyBuf)
+			c.Assert(joinHash&radixBits, Equals, uint32(i)&radixBits)
+		}
+	}
+	c.Assert(totalRowCnt, Equals, 200)
+
+	for _, ch := range hashJoinExec.outerResultChs {
+		close(ch)
+	}
+	for _, ch := range hashJoinExec.joinChkResourceCh {
+		close(ch)
+	}
+	err = hashJoinExec.Close()
+	c.Assert(err, IsNil)
+}
+
```

Comment:
This unit test would need update once we modify lines in `fetchInnerAndBuildHashTable` or its callees? seems it is pretty fragile.

Reply:
Seems not, we only rely on the data of HashJoinExec.innerResult, but not the construct procedure for it.
But if the process of radix-partition is modified, we also need to change this test.